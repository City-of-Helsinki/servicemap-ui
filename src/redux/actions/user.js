import LocalStorageUtility from '../../utils/localStorage';
import fetchAddress from '../../views/MapView/utils/fetchAddress';

export const setLocale = locale => ({
  type: 'SET_LOCALE',
  locale,
});

export const setInitialLoad = () => ({
  type: 'SET_INITIAL_LOAD',
});

export const setPage = page => ({
  type: 'SET_CURRENT_PAGE',
  page,
});

export const setUserPosition = position => ({
  type: 'SET_POSITION',
  position,
});

const setTheme = theme => ({
  type: 'SET_THEME',
  theme,
});


export const changeLocaleAction = locale => async (dispatch) => {
  dispatch(setLocale(locale));
};

export const actionSetInitialLoad = () => async (dispatch) => {
  dispatch(setInitialLoad());
};

export const setCurrentPage = page => async (dispatch) => {
  dispatch(setPage(page));
};

export const changeTheme = theme => async (dispatch) => {
  LocalStorageUtility.saveItem('theme', theme);
  dispatch(setTheme(theme));
};

export const findUserLocation = () => async (dispatch) => {
  const success = (position) => {
    if (position.coords.accuracy < 10000) {
      fetchAddress({ lat: position.coords.latitude, lng: position.coords.longitude })
        .then((data) => {
          dispatch(setUserPosition({
            coordinates: position.coords,
            allowed: true,
            addressData: data,
          }));
        });
    } else {
      console.warn(`Position accuracy: ${position.coords.accuracy}. Max accuracy: 10000`);
      dispatch(setUserPosition({ coordinates: null, allowed: true, addressData: null }));
    }
  };

  const error = (err) => {
    console.warn(`GeoLocation error:(${err.code}): ${err.message}`);
    if (err.code === 1) {
      dispatch(setUserPosition({ coordinates: null, allowed: false, addressData: null }));
    } else {
      dispatch(setUserPosition({ coordinates: null, allowed: true, addressData: null }));
    }
  };

  navigator.geolocation.getCurrentPosition(success, error);
};


export default { setLocale, changeLocaleAction };
