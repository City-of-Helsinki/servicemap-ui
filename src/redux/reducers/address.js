const initialState = {
  units: [],
  addressCoordinates: null,
  addressData: null,
  adminDistricts: [],
  toRender: 'adminDistricts',
};

export default (state = initialState, action) => {
  switch (action.type) {
    case 'SET_ADDRESS_DATA':
      return {
        ...state,
        addressData: action.data,
      };
    case 'SET_ADDRESS_LOCATION':
      return {
        ...state,
        addressCoordinates: action.location.addressCoordinates,
      };
    case 'SET_ADDRESS_UNITS':
      return {
        ...state,
        units: action.units,
      };
    case 'SET_ADMINISTRATIVE_DISTRICTS':
      return {
        ...state,
        adminDistricts: action.data,
      };
    case 'SET_TO_RENDER':
      return {
        ...state,
        toRender: action.data,
      };
    default:
      return state;
  }
};
