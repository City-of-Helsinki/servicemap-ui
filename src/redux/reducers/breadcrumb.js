const initialState = [];

const breadcrumbReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'PUSH_ENTRY':
      return [...state, action.entry];
    case 'POP_ENTRY': {
      const newEntries = Array.from(state);
      newEntries.pop();
      return newEntries;
    }
    case 'REPLACE_ENTRY': {
      const newEntries = Array.from(state);
      newEntries.pop();
      return [...newEntries, action.entry];
    }
    default:
      return state;
  }
};

export default breadcrumbReducer;
