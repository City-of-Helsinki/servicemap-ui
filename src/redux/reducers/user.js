
const initialState = {
  initialLoad: false,
  locale: 'fi',
  page: 'home',
  theme: 'default',
  position: {
    coordinates: null,
    allowed: false,
  },
};

export default (state = initialState, action) => {
  switch (action.type) {
    case 'SET_LOCALE':
      return {
        ...state,
        locale: action.locale,
      };
    case 'SET_INITIAL_LOAD':
      return {
        ...state,
        initialLoad: true,
      };
    case 'SET_CURRENT_PAGE':
      return {
        ...state,
        page: action.page,
      };
    case 'SET_POSITION':
      return {
        ...state,
        position: action.position,
      };
    case 'SET_THEME':
      return {
        ...state,
        theme: action.theme,
      };
    default:
      return state;
  }
};
