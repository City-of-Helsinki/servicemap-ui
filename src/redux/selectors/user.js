
export const getTheme = state => state.user.theme;
export const getPage = state => state.user.page;
export const getLocale = state => state.user.locale;
