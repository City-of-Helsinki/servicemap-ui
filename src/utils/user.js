import { useSelector } from 'react-redux'
 
export const useUserTheme = () => useSelector(state => state.user.theme);
export const useUserLocale = () => useSelector(state => state.user.locale);
