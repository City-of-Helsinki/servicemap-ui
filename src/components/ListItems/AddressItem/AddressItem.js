import React from 'react';
import PropTypes from 'prop-types';
import { uppercaseFirst } from '../../../utils';
import SimpleListItem from '../SimpleListItem';
import { AddressIcon } from '../../SMIcon';

const AddressItem = (props) => {
  const {
    getLocaleText, navigator, address, classes,
  } = props;

  const number = `${address.number ? address.number : ''}${address.letter ? address.letter : ''}`;
  const text = `${getLocaleText(address.street.name)} ${number}, ${uppercaseFirst(address.street.municipality)}`;

  return (
    <SimpleListItem
      button
      text={uppercaseFirst(text)}
      icon={<AddressIcon className={classes.icon} />}
      divider
      handleItemClick={(e) => {
        e.preventDefault();
        if (navigator) {
          navigator.push('address', {
            municipality: address.street.municipality,
            street: getLocaleText(address.street.name),
            number: `${number}`,
          });
        }
      }}
      role="link"
    />
  );
};


export default AddressItem;

AddressItem.propTypes = {
  classes: PropTypes.objectOf(PropTypes.any).isRequired,
  getLocaleText: PropTypes.func.isRequired,
  navigator: PropTypes.objectOf(PropTypes.any),
  address: PropTypes.objectOf(PropTypes.any).isRequired,
};

AddressItem.defaultProps = {
  navigator: null,
};
