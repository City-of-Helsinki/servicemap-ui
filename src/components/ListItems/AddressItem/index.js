import AddressItem from './AddressItem';

export default AddressItem;
