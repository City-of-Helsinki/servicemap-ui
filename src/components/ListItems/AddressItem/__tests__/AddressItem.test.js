import React from 'react';
import { initialState } from '../../../../redux/reducers/user';
import AddressItem from '../index';
import { getRenderWithProviders } from '../../../../../jestUtils';

const mockData = {
  "name": {
      "fi": "yleiset kirjastot",
      "sv": "allm√§nna bibliotek",
      "en": "public libraries"
  },
  "id": 813,
  "clarification_enabled": false,
};

// Generic required props for SimpleListItem
const mockProps = {
  service: mockData,
};


const renderWithProviders = getRenderWithProviders({
  user: initialState,
  settings: {},
  service: {
    current: null,
  },
})

describe('<AddressItem />', () => {
  it('should work', () => {
    const { container } = renderWithProviders(<AddressItem {...mockProps} />);
    expect(container).toMatchSnapshot();
  });

  it('does render accessibility attributes correctly', () => {
    const { container } = renderWithProviders(<AddressItem {...mockProps} />);
    const items = container.querySelectorAll('li');
    const firstItem = items[0];
    const firstItemResultTitle = firstItem.querySelectorAll('p')[0];
    const dividerItem = items[1];

    // List item's image should be aria-hidden
    expect(firstItem.querySelector('span').getAttribute('aria-hidden')).toBeTruthy();
    expect(firstItem.getAttribute('role')).toEqual('link');
    expect(firstItem.getAttribute('tabIndex')).toEqual('0');
    expect(firstItemResultTitle.getAttribute('aria-hidden')).toBeFalsy();
    expect(dividerItem.getAttribute('aria-hidden')).toBeTruthy();
  });
});
