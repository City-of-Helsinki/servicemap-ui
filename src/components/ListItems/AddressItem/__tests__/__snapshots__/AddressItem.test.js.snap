// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddressItem /> should work 1`] = `
<div>
  <li
    class="MuiButtonBase-root SimpleListItem-listItem-2 MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-button null  css-16e6fe9-MuiButtonBase-root-MuiListItem-root"
    role="link"
    tabindex="0"
  >
    <div
      aria-hidden="true"
      class="MuiListItemIcon-root SimpleListItem-listIcon-6  css-cveggr-MuiListItemIcon-root"
    >
      <span
        aria-hidden="true"
        class="Connect(AddressItem)-icon-1 icon-icon-address SMIcon-icon-9"
      />
    </div>
    <div
      class="SimpleListItem-textContainer-3 MuiListItemText-root css-vt7cuc-MuiListItemText-root"
    >
      <p
        class="null  MuiTypography-root MuiTypography-body2 css-q91igz-MuiTypography-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 css-1ak20dk-MuiTypography-root"
          style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 1px;"
        />
        
      </p>
    </div>
  </li>
  <li
    aria-hidden="true"
  >
    <hr
      class="MuiDivider-root MuiDivider-fullWidth SimpleListItem-divider-7 css-9mgopn-MuiDivider-root"
    />
  </li>
</div>
`;
